### ISSUE TITLE
“include” instructions in Compose files should support %L and %H (was: Help with setup on Fedora 31)

### ISSUE THREAD

**Issue by akiross** [mike-fabian/ibus-typing-booster#93]
Hello, I wanted to give typing booster a try on Fedora 31 using my regular keyboard layout + typing booster to enter unicode symbols quickly. I tried following the doc on the github page, but it doesn't seem to be quite up to date with my environment.

So, here's what I did (using gnome 3.34.5)

 - installed via dnf: `dnf install -y ibus-typing-booster` -> already installed
 - started ibus-daemon -> `current session already has an ibus-daemon.`
 - went to Gnome Settings -> Region & Languages -> Added `Other (Typing Booster)` to my input sources
 - Opened settings, added a couple of dictionaries (en_US and it_IT) and installed hunspell dictionaries
 - Checked the option: "unicode symbols and emoji prediction" (all other options are off)
 - Closed the settings
 - In the GNOME top bar, I switched to Typing Booster input (edit: I also tried with my regular input)
 - Opened a new window (gedit) and started typing
 - Nothing pops up

I never used ibus before, so I'm not really sure I configured things correctly.
Can you please help me in getting TB working?

**Comment by mike-fabian** [comment]
That is really weird, something should pop up at that moment.

Is there something interesting in the log file?:
``
~/.local/share/ibus-typing-booster/debug.log
``

**Comment by akiross** [comment]
Ok, the log was useful because apparently I was missing an error message: my `.XCompose` starts with the line `include "%L"` but if I run `ibus-daemon -rR`, one error pops up:

```
Traceback (most recent call last):
  File "/usr/share/ibus-typing-booster/engine/itb_util.py", line 3243, in _read_compose_file
    errors='ignore') as compose_file:
FileNotFoundError: [Errno 2] No such file or directory: '%L'
```

which basically stopped everything. After noticing this, I removed that line from .XCompose and it works as expected.

Thank you!

**Comment by mike-fabian** [comment]
Where did this "%L" come from? Does %L have a special meaning?

I parse regular file names in that place, i.e. you could use something like
``
include "/usr/share/X11/locale/en_US.UTF-8/Compose"
``
there if you want (although that English compose file is included automatically if you are running in en_US.UTF-8 locale). But you could include any file you like there if it is valid Compose  file.

If %L has a special meaning I didn’t know of, I should handle it as well.

**Comment by mike-fabian** [comment]
https://wiki.debian.org/XCompose says:

The second step is to configure your own custom compose sequences. To do this, create a file named ~/.XCompose and make sure it has include "%L" at the top of it, to bring in the default sequences. After that, you may put whatever you like. For example,
``
include "%L"

<Multi_key> <g> <a> : "α"
<Multi_key> <g> <b> : "β"
<Multi_key> <g> <g> : "γ"
``

**Comment by mike-fabian** [comment]
https://www.x.org/releases/X11R7.5/doc/man/man5/Compose.5.html

says:

 Compose files can use an "include" instruction. This allows local modifications to be made to existing compose files without including all of the content directly. For example, the system's iso8859-1 compose file can be included with a line like this:

    include "/usr/share/X11/locale/iso8859-1/Compose" 

There are two substitutions that can be made in the file name of the include instruction. %H expands to the user's home directory (the $HOME environment variable), and %L expands to the name of the locale specific Compose file (i.e., "/usr/share/X11/locale/<localename>/Compose").

For example, you can include in your compose file the default Compose file by using:

    include "%L" 

and then rewrite only the few rules that you need to change. New compose rules can be added, and previous ones replaced.

**Comment by akiross** [comment]
Ah yes, I suspected this was a bug, but didn't want to use the same discussion. Meanwhile, having it working I got some questions regarding the configuration of typing booster. Not sure if it's ok to use this issue, but:

 - is it possible to completely disable dictionary based suggestions? If I remove all dictionaries, I still end up having the english one in the list
 - is it possible to have a custom list of entries? Something like XCompose, where I can associate a work to a symbol and have it suggested. It is my understanding that [XCompose is handled somehow](https://github.com/mike-fabian/ibus-typing-booster/blob/e14068cd4d6275f9395043a86b9eff00baa9b0d6/engine/itb_util.py#L3000) but it is not clear how could I enable *just* that table (or an alternative one).

**Comment by akiross** [comment]
Also, watching the code I just noticed that, in the `ComposeSequence` class, conflicting sequences are resolved with just the last sequence being saved. I am looking for custom tables that allow to have conflicting entries.

**Comment by mike-fabian** [comment]
> Also, watching the code I just noticed that, in the `ComposeSequence` class, conflicting sequences are resolved with just the last sequence being saved. 

According to 

https://www.x.org/releases/X11R7.5/doc/man/man5/Compose.5.html

That is how compose tables work, later entries override previous ones.

I am already following the specification not 100% because I read the default compose file automatically, even when it is not included in `~/.XCompose` and I also read `'~/.config/ibus/Compose'` (with lower priority than `~/.XCompose` but higher priority than the system default Compose file). `'~/.config/ibus/Compose'` is not specified in the above man-page, but ibus uses it, so I read that one as well. On top of that, I try to interpret any dead key sequence and produce a reasonable result even if no such dead key sequence is specified in any compose file. 

> I am looking for custom tables that allow to have conflicting entries.

I think the Compose mechanism would not work well with conflicting entries.

But maybe the tab in the setup tool where you can define custom shortcuts does what you want?

There you could define something like

smile ☺

smile ☹

which conflict, and if you type “smile” you get both ☺ and ☹ as high priority candidates.

**Comment by mike-fabian** [comment]
> is it possible to completely disable dictionary based suggestions? If I remove all dictionaries, I still end up having the english one in the list

At the moment this is not possible, I didn’t think this was useful. Therefore, as you noticed, if one tries to make the list of dictionaries empty, the US English dictionary is automatically added back as a default. 

I think it would be not very difficult to allow deleting all dictionaries.

But why would anyone want to do that?

**Comment by akiross** [comment]
> But maybe the tab in the setup tool where you can define custom shortcuts does what you want?

Custom shortcuts are what I was needing, thanks. Is there a file I can directly edit without using the GUI?

> I think it would be not very difficult to allow deleting all dictionaries.
> But why would anyone want to do that?

Well, in my case because I prefer to use specific spell checking when writing (e.g. firefox's) and having a dictionary slows down my usage of typing booster, since I just need it for symbols.

I'm heavily using XCompose, which is great, but is very bad in case of conflicts, and it is not as extensible as typing booster. In particular, my desired interaction is the following:

 - if I have a custom shortcut, that should pop up first when matching.
 - if there is no match, search unicode/emoji table.

Dictionaries are just introducing noise in symbol search (also, on my slow computer they are kinda of slowing down the typing), so I'd rather avoid them.

This being said, it could be that I'm just mis-using typing booster :)

I noticed there are very advanced things (such as learning from what is being typed, suggestions based on NLTK and so on), but I prefer start simple, with the bare minimum, and enable more features as I go.

**Comment by mike-fabian** [comment]
> Custom shortcuts are what I was needing, thanks. Is there a file I can directly edit without using the GUI? (**EDIT:** This is because I tend to share dotfiles among hosts and sharing typing booster configuration would be wonderful as well.)

No, these custom shortcuts are stored in the user database where the data learned from typing is stored as well.

Attention: This means that if you click on the button “Delete learned data” in the setup tool, you will loose all custom shortcuts as well!!!

For example, I have defined a custom shortcut “mfg“ to expand as “Mit freundlichen Grüßen”. This has been added to my user database as if I had typed “mfg” and selected a candidate “Mit freundlichen Grüßen” one million times:

``
$ sqlite3 ~/.local/share/ibus-typing-booster/user.db
SQLite version 3.30.0 2019-10-04 15:03:17
Enter ".help" for usage hints.
sqlite> select * from phrases where input_phrase='mfg';
318110|mfg|Mit freundlichen Grüßen|||1000002|1545519220.42398
sqlite> 
``

Actually you see 1000002 as the user frequency there because I have used this custom shortcut 2 times after adding it (Just did that a few minutes ago for testing).

The one million is just an arbitrary high number, high enough that it is almost impossible to reach by learning from normal typing.

As it is almost impossible to reach a count of one million by just typing the same stuff and selecting the same candidate over and over again and custom shortcuts have a count of at least one million, they will always be very high up in the candidate list, higher than all candidates which have been learned from regular typing.

If you define conflicting custom shortcuts like

smile ☺

smile ☹

the one which you use more often will go to the top of the candidate list because each time you use it the count increases by one.

About the dot file sharing between hosts:

Currently, the only way to get the custom shortcuts and all the other stuff which has been recorded in the user database is to copy the binary user.db to the other host.

To make copying the user database to a different host easier, and to enable easier manipulation with a text editor, I have this idea on my TODO list for a long time already:

Add an options to

- export the contents of the user database to a plain text file

- read such a plain text file and merge it into the existing user.db

- read such a plain text file and throw away the existing userdb and create a new one from the text file

That would also make it easier to make a backup of the user.db. At the moment I just copy the binary to a backup file from time to time, just in case I accidentally delete the user.db by testing the “Delete learned data” option in the setup tool.

**Comment by mike-fabian** [comment]
> Dictionaries are just introducing noise in symbol search (also, on my slow computer they are kinda of slowing down the typing), so I'd rather avoid them.

Can you create a new issue “Allow to delete *all* dictionaries in the setup tool” so I don’t forget this?

I think it is not difficult to do, but it is better to have a seperate issue so I don’t forget it.

**Comment by mike-fabian** [comment]
The fix is included in the 2.8.3 release:

https://github.com/mike-fabian/ibus-typing-booster/releases/tag/2.8.3
